### ISSUE TITLE
Explicit repository credentials break with configuration caching

### ISSUE THREAD

**Issue by ccjernigan** [gradle/gradle#24040]
### Expected Behavior
Explicitly set credentials for a Maven repository are used.

### Current Behavior
Build fails, asking for implicit properties to be configured.  E.g.
```
* What went wrong:
Execution failed for task ':publish:publishDataPublicationToGitHubPackageRegistryRepository'.
> The following Gradle properties are missing for 'GitHubPackageRegistry' credentials:
    - GitHubPackageRegistryUsername
    - GitHubPackageRegistryPassword
```

### Context
I have a simple Gradle project to publish dataset JAR files to GitHub Package Registry, as seen in the sample below. (The datasets are effectively a ZIP of a directory, so this is a reasonable way to get them uploaded to be consumed by another project behind authentication).

This fails in Gradle 8.0.1.

### Steps to Reproduce 
1. Set up a script with the code below
1. Generate a GitHub Personal Access token
    1. Generate a new [Personal Access Token](https://github.com/settings/tokens/new)
    1. Put in a name for the token, such as `package registry`
    1. Check the box for `write:packages` only
    1. Click Generate Token
    1. Copy the token to be used in the next step
1. Enable configuration caching `org.gradle.unsafe.configuration-cache=true` in gradle.properties
1. Run the Gradle command, supplying the GitHub Package Registry details (URL, user, and token) and the package details (group, artifact id, version).  This command would look something like: `./gradlew publishAllPublicationsToGitHubPackageRegistryRepository -PPUBLISHER_GITHUB_PACKAGE_REGISTRY_URL= -PPUBLISHER_GITHUB_PACKAGE_REGISTRY_USER= -PPUBLISHER_GITHUB_PACKAGE_REGISTRY_TOKEN= -PPUBLISHER_GROUP_ID= -PPUBLISHER_ARTIFACT_ID= -PPUBLISHER_VERSION=`

In build.gradle.kts
```
plugins {
    id("maven-publish")
}

publishing {
    repositories {
        mavenLocal {
            name = "MavenLocal"
        }

        val githubPackageRegistryUrl = extra["PUBLISHER_GITHUB_PACKAGE_REGISTRY_URL"].toString()
        val githubPackageRegistryUsername = extra["PUBLISHER_GITHUB_PACKAGE_REGISTRY_USER"].toString()
        val githubPackageRegistryToken = extra["PUBLISHER_GITHUB_PACKAGE_REGISTRY_TOKEN"].toString()
        if (githubPackageRegistryUrl.isNotBlank() && githubPackageRegistryUsername.isNotBlank() && githubPackageRegistryToken.isNotBlank()) {
            maven(githubPackageRegistryUrl) {
                name = "GitHubPackageRegistry"
                credentials {
                    username = githubPackageRegistryUsername
                    password = githubPackageRegistryToken
                }
            }
        }
    }

    publications {
        create<MavenPublication>("data") {
            groupId = extra["PUBLISHER_GROUP_ID"].toString()
            artifactId = extra["PUBLISHER_ARTIFACT_ID"].toString()
            version = extra["PUBLISHER_VERSION"].toString()

            // Place a file with this name in the root directory
            artifact(File(rootDir, "to_publish.jar"))
        }
    }
}

```

### Your Environment

Build scan URL:

**Comment by ccjernigan** [comment]
I previously flagged this as a regression, but I don't think that's true because I didn't have the configuration cache enabled when I was using Gradle 7.6.  There were several moving parts to check when trying to get the reproduction steps narrowed down.

**Comment by jbartok** [comment]
Thank you for your interest in Gradle!

This issue needs a decision from the team responsible for that area. They have been informed, response time may vary.

**Comment by abstratt** [comment]
@ccjernigan Our documentation/implementation is at a messy state regarding this.

Up to 7.5, there was no support for Maven/Ivy publishing with the configuration cache. In 7.6.1/8.0, that support was implemented, but the choice then was not to persist credentials (at first we had an error message letting you know inlined/explicit credentials were not supported, but that was later reverted for being too harsh and still [not fully effective](https://github.com/gradle/gradle/issues/22618#issuecomment-1304865537)).

For 8.1, we are considering supporting [encryption of the configuration cache state files](https://docs.google.com/document/d/18UISyLGyrC_3lhiM4ZygTvFp4f6wt7scdgyuRhPObKk/edit), and while doing so, changing the persistence of Maven/Ivy repos in the Maven/Ivy Publish plugin configurations so inlined/explicit credentials are preserved, but that is in the air now.

Meanwhile, just so you are unblocked, is there a technical reason why you could not externalize those credentials, as [shown here](https://docs.gradle.org/current/samples/sample_publishing_credentials.html#header)?

**Comment by efemoney** [comment]
Adding our own use case here. 

We are creating a custom distribution that includes an init script to further simplify the setup for our engineers!

Today its cumbersome because to use our **internal plugins** served from our **internal artifactory**, the exact same block of code needs to be specified in
- `PluginManagement.repositories`
- `DependencyResolutionManagement.repositories`
- (possibly every single) `Project.repositories`
- (every single) Included Build `Project.repositories`

We are adding an init script that for each added repository (`RepositoryHandler.configureEach`) across these different points, we automatically add the correct credentials 

```kotlin
username = providers.gradleProperty(...).orElse(providers.environmentVariable(...))
password = providers.gradleProperty(...).orElse(providers.environmentVariable(...))
```

Didn't even get to test the init script because this is failing with this issue at publishing the custom distribution using the ivy-publish plugin 😅 

Based on the scenario above it is clear that **we will not know** the names of these added repositories before hand and even if we do there is no way to influence the loaded gradle properties so that we could have `repositoryNameUsername` & `repositoryNamePassword` gradle properties.

**Comment by cusxy** [comment]
I encountered the same problem.

I see that this occurs when I set the `org.gradle.configuration-cache.problems` property to `warn`. When I use the default `fail` strategy, the `maven-plugin` works correctly.

It seems that this leads to unexpected behavior with the `Cached` class. When `PublishToMavenRepository::publish` starts, it tries to retrieve the `spec` variable of `Cached<PublishSpec>` class.

* With the `fail` strategy, the plugin correctly gets a `repository` object of the `RepositorySpec.Configured` class.
* however, with the `warn` strategy, the plugin gets an object of the `RepositorySpec.DefaultRepositorySpec` class, with seems unexpected (see `::computeSpec` fun).

I suspect that this could potentially lead to other errors when using the configuration cache together with the `problem` property. _And, maybe, I should create another issue to describe that specific error._

To demonstrate this issue, I created a simple project. You can reproduce the error by running the following command:

```sh
./gradlew :lib:publishReleasePublicationToReleaseRepository
```

If you set `org.gradle.configuration-cache.problems` to `fail` in the `gradle.properties` file, you will see a different error. However, that error is expected because the project uses a made-up URL.

The example project: https://github.com/cusxy/configuration-cache-problems

**Comment by abstratt** [comment]
Update for Gradle 9

In Gradle 9, the observed behavior has changed: now, when explicit credentials are used, instead of causing a build failure, explicit credentials disable the configuration cache.
